src/Hive.IO/Hive.IO/EnergySystems/CombustionTech.cs

```diff
@@ -24,33 +22,34 @@ public GasBoiler(double investmentCost, double embodiedGhg, double capacity, dou
         }
 
 
-        protected override double[] SetConversionEfficiencyHeating()
-        {
-            throw new System.NotImplementedException();
-        }
-
-
-        //where is gas defined?? for solar, its clear, its from a weather file...
-        // so all Input Carriers should be part of Hive.IO.Environment!
-        // indicate, whether we have biogas, wood pellets, district heating, electricity grid, natural gas, oil, etc
-        public void SetInput(Gas gasInput)
+        public void SetInputOutput(Gas gasInput, double[] heatingGenerated, double[] supplyTemperature)
         {
-            base.InputCarrier = gasInput;
-        }
+            int horizon = heatingGenerated.Length;
+
+            var gasConsumed = new double[horizon];
+            var gasPrice = new double[horizon];
+            var gasEmissionsFactor = new double[horizon];
+
+            for (int t = 0; t < horizon; t++)
+            {
+                gasConsumed[t] = heatingGenerated[t] / this.Efficiency;
+            }
+            if (horizon == Misc.MonthsPerYear)
+            {
+                gasPrice = Misc.GetAverageMonthlyValue(gasInput.EnergyPrice);
```

Comment:
`Misc.GetAverageMonthlyValue` could probably also be written in functional style as something like (I'd have to look up the exact method names and how they're used, but the principle is the same):

`gasPrice = gasInput.EnergyPrice.GroupBy(p => Month(p)).Select(group => group.Average())`

of course, here it might be a bit harder to come up with `Month` and `Average` but you get the idea?